onnxruntime/core/optimizer/bias_softmax_fusion.cc

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/logging/logging.h"
+#include "core/providers/common.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/bias_softmax_fusion.h"
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/utils.h"
+#include <deque>
+
+#define ILOGF(format_str, ...) \
+  LOGF_DEFAULT(INFO, format_str, ##__VA_ARGS__)
+
+using namespace ONNX_NAMESPACE;
+using namespace ::onnxruntime::common;
+namespace onnxruntime {
+
+// helper to check dimensions match on concrete or symbolic value
+bool operator!=(
+  const ONNX_NAMESPACE::TensorShapeProto_Dimension& lhs,
+  const ONNX_NAMESPACE::TensorShapeProto_Dimension& rhs) {
+    return !(lhs == rhs);
```

Comment:
does TensorShapeProto_Dimension only have operator==() defined? that seems odd. out of curiosity, where is that coming from?

Reply:
Moved to onnxruntime/core/framework/tensorprotoutils.cc.